File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodThrowsVerifier.java
package com.jetbrains.pluginverifier.verifiers.method;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.api.VContext;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.utils.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;

import java.util.List;

/**
 * @author Sergey Patrikeev
 */
public class MethodThrowsVerifier implements MethodVerifier {
  @SuppressWarnings("unchecked")
  @Override
  public void verify(ClassNode clazz, MethodNode method, Resolver resolver, VContext ctx) {
    List<String> exceptions = (List<String>) method.exceptions;
    for (String exception : exceptions) {
      String descr = VerifierUtil.extractClassNameFromDescr(exception);
      if (descr == null) continue;
      if (!VerifierUtil.classExistsOrExternal(ctx, resolver, descr)) {
        ctx.registerProblem(new ClassNotFoundProblem(descr), ProblemLocation.fromMethod(clazz.name, method));
      }
    }
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodThrowsVerifier.kt
package com.jetbrains.pluginverifier.verifiers.method

import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescriptor
import com.jetbrains.pluginverifier.verifiers.resolution.Method

class MethodThrowsVerifier : MethodVerifier {
  override fun verify(method: Method, context: VerificationContext) {
    for (exception in method.exceptions) {
      val className = exception.extractClassNameFromDescriptor() ?: continue
      context.classResolver.resolveClassChecked(className, method, context)
    }
  }
}
